src/core/struct_tree.js

```diff
@@ -67,6 +69,425 @@ class StructTreeRoot {
+
+        buffer.length = 0;
+        await writeObject(tagRef, tagDict, buffer, xref);
+        newRefs.push({ ref: tagRef, data: buffer.join("") });
+
+        nums.push(parentTreeId, tagRef);
+        kids.push(tagRef);
+      }
+    }
+    return nextKey + 1;
+  }
+
+  static #collectParents({ elements, xref, pageDict, parentTree }) {
+    const idToElement = new Map();
+    for (const element of elements) {
+      if (element.structTreeParentId) {
+        const id = parseInt(element.structTreeParentId.split("_mc")[1], 10);
+        idToElement.set(id, element);
+      }
+    }
+
+    const id = pageDict.get("StructParents");
+    const numberTree = new NumberTree(parentTree, xref);
+    const parentArray = numberTree.get(id);
+    if (!Array.isArray(parentArray)) {
+      return;
+    }
+    const updateElement = (kid, pageKid, kidRef) => {
+      const element = idToElement.get(kid);
+      if (element) {
+        const parentRef = pageKid.getRaw("P");
+        const parentDict = xref.fetchIfRef(parentRef);
+        if (parentRef instanceof Ref && parentDict instanceof Dict) {
+          // It should always the case, but we check just in case.
+          element.structTreeParent = { ref: kidRef, dict: pageKid };
+        }
+        return true;
+      }
+      return false;
+    };
+    for (const kidRef of parentArray) {
+      if (!(kidRef instanceof Ref)) {
+        continue;
+      }
+      const pageKid = xref.fetch(kidRef);
+      const k = pageKid.get("K");
+      if (Number.isInteger(k)) {
+        updateElement(k, pageKid, kidRef);
+        continue;
+      }
+
+      if (!Array.isArray(k)) {
+        continue;
+      }
+      for (let kid of k) {
+        kid = xref.fetchIfRef(kid);
+        if (Number.isInteger(kid) && updateElement(kid, pageKid, kidRef)) {
+          break;
+        }
+      }
+    }
+  }
+
+  static async #updateParentTag({
+    structTreeParent: { ref, dict },
+    tagDict,
+    newTagRef,
+    xref,
+    newRefs,
+    buffer,
+  }) {
+    // We get the parent of the tag.
+    const parentRef = dict.getRaw("P");
+    let parentDict = xref.fetchIfRef(parentRef);
+
+    tagDict.set("P", parentRef);
+
+    // We get the kids in order to insert a new tag at the right position.
+    let saveParentDict = false;
+    let parentKids;
+    let parentKidsRef = parentDict.getRaw("K");
+    if (!(parentKidsRef instanceof Ref)) {
+      parentKids = parentKidsRef;
+      parentKidsRef = xref.getNewTemporaryRef();
+      parentDict = parentDict.clone();
+      parentDict.set("K", parentKidsRef);
+      saveParentDict = true;
+    } else {
+      parentKids = xref.fetch(parentKidsRef);
+    }
+
+    if (Array.isArray(parentKids)) {
+      const index = parentKids.indexOf(ref);
+      if (index >= 0) {
+        parentKids = parentKids.slice();
+        parentKids.splice(index + 1, 0, newTagRef);
+      } else {
+        warn("Cannot update the struct tree: parent kid not found.");
+        return;
```

Comment:
Given that this is only an early return, and thus won't stop further parsing, what's the impact of hitting this branch?
It seems, if I'm following the code correctly, that the "rest" of the structTree updating still continues in this case and it's not immediately clear to me what happens then?